query-algebrizer/src/clauses/not.rs

```diff
@@ -177,22 +170,64 @@ mod testing {
         let mut subquery = ConjoiningClauses::default();
         subquery.from.append(&mut vec![SourceAlias(DatomsTable::Datoms, d1),
                                   SourceAlias(DatomsTable::Datoms, d2)]);
-        subquery.column_bindings.insert(vx.clone(), vec![d1e.clone(), d2e.clone()]);
+        subquery.column_bindings.insert(vx.clone(), vec![d0e.clone(), d1e.clone(), d2e.clone()]);
         subquery.wheres.add(ColumnConstraintOrAlternation::Constraint(ColumnConstraint::Equals(d1a.clone(), parent)));
         subquery.wheres.add(ColumnConstraintOrAlternation::Constraint(ColumnConstraint::Equals(d1v.clone(), ambar)));
-        subquery.wheres.add(ColumnConstraintOrAlternation::Constraint(ColumnConstraint::Equals(d2a.clone(), subquery_knows)));
+        subquery.wheres.add(ColumnConstraintOrAlternation::Constraint(ColumnConstraint::Equals(d2a.clone(), knows.clone())));
         subquery.wheres.add(ColumnConstraintOrAlternation::Constraint(ColumnConstraint::Equals(d2v.clone(), daphne)));
-        subquery.wheres.add(ColumnConstraintOrAlternation::Constraint(ColumnConstraint::Equals(d1e.clone(), QueryValue::Column(d2e.clone()))));
+        subquery.wheres.add(ColumnConstraintOrAlternation::Constraint(ColumnConstraint::Equals(d0e.clone(), QueryValue::Column(d1e.clone()))));
+        subquery.wheres.add(ColumnConstraintOrAlternation::Constraint(ColumnConstraint::Equals(d0e.clone(), QueryValue::Column(d2e.clone()))));
 
-        let mut s = HashSet::with_capacity(1);
-        s.insert(ValueType::Ref);
-        subquery.known_types.insert(vx.clone(), s);
+        subquery.known_types.insert(vx.clone(), unit_type_set(ValueType::Ref));
 
         assert!(!cc.is_known_empty());
         assert_eq!(cc.wheres, ColumnIntersection(vec![
                 ColumnConstraintOrAlternation::Constraint(ColumnConstraint::Equals(d0a.clone(), knows.clone())),
                 ColumnConstraintOrAlternation::Constraint(ColumnConstraint::Equals(d0v.clone(), john)),
-                ColumnConstraintOrAlternation::Constraint(ColumnConstraint::NotExists(ComputedTable::Subquery{ clauses: subquery })),
+                ColumnConstraintOrAlternation::Constraint(ColumnConstraint::NotExists(ComputedTable::Subquery(subquery))),
+            ]));
+        assert_eq!(cc.column_bindings.get(&vx), Some(&vec![d0e]));
+        assert_eq!(cc.from, vec![SourceAlias(DatomsTable::Datoms, d0)]);
+    }
+
+    // not-join.
+    #[test]
+    fn test_successful_not_join() {
+        let schema = prepopulated_schema();
+        let query = r#"
+            [:find ?x
+             :where [?x :foo/knows ?y]
+                    (not-join [?x] 
+                              [?x :foo/parent ?y])]"#;
```

Comment:
If you extend this query to

```
[:find ?x
 :where
 [?x :foo/knows ?y]
 [?x :foo/age 11]
 [?x :foo/name "Bill"]
 (not-join [?x]
   [?x :foo/parent ?y])]
```

you should see what I mean about expanding column bindings inside the child.